File: 1143-longest-common-subsequence/1143-longest-common-subsequence.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        dp = [0 for _ in range(len(text1))]
        
        for i in range(len(text2)):
            for j in range(len(dp)-1,-1,-1):
                if text1[j] == text2[i]:
                    if j == 0:
                        dp[j] = 1
                    else:
                        dp[j] = max(dp[j],max(dp[:j]) + 1)
            # print(dp)
        return max(dp)

File: 1143-longest-common-subsequence/1143-longest-common-subsequence.java
class Solution {
    public int longestCommonSubsequence(String text1, String text2) {
        int[] dp = new int[text1.length()];
        int ans = 0;
        for (int i = 0 ; i < text2.length();i ++){
            for (int j = dp.length -1 ; j >= 0 ; j --){
                if (text1.charAt(j) == text2.charAt(i)){
                    int maxVal = 0;
                    for (int k = 0 ; k < j ; k ++){
                        maxVal = Math.max(dp[k],maxVal);
                    }
                    dp[j] = Math.max(dp[j],maxVal + 1);
                    ans = Math.max(dp[j],ans);
                }
            }
        }
        return ans;
    }
}